### PR TITLE
[[FEAT]] parse and lint tagged template literals

### LINKED ISSUE #2000: Tagged Template Strings are not linting correctly
I'm getting a ```Missing semicolon.``` error when using a tagged template string. If I have time I'll try to submit a test case for this soon.

My code looks roughly like:
```
function tag() {}

var testStr = 'bar';
function foo() {
  return tag `some
    ${testStr} here`;
}
foo();

```

### PR DESCRIPTION
**Pull request by caitp**
Closes #2000

### CODE REVIEW

**Comment by caitp** [comment]
This was really hard to make work, the pratt parser is danged confusing =) anyways, there's probably a lot of gunk in this patch that isn't really needed, so lets try to figure out what can be safely removed before merging

File: src/jshint.js [hunk 1]
```diff
@@ -1490,7 +1488,19 @@ var JSHINT = (function() {
   }
 
   function parseFinalSemicolon() {
+    if (lex.inContext(Context.Template)) return;
```
Comment (caitp) [inline src/jshint.js]: I'm not sure if this is really needed... technically template expressions aren't parsed using statement() so they shouldn't need this...

File: src/jshint.js [hunk 2]
```diff
@@ -1490,7 +1488,19 @@ var JSHINT = (function() {
...
     if (state.tokens.next.id !== ";") {
+      if (state.tokens.next.isUnclosed) {
+        advance();
+        return;
+      }
+      if (state.tokens.next.id === "(end)") {
+        // If we got here via an unclosed template literal, don't complain.
+        if ((state.tokens.curr.template && state.tokens.curr.isUnclosed) ||
```
Comment (caitp) [inline src/jshint.js]: these could probably be organized better ._.

File: src/jshint.js [hunk 3]
```diff
@@ -1985,24 +1996,37 @@ var JSHINT = (function() {
...
-
-  type("(no subst template)", function() {
-    return this;
-  });
+  state.syntax["(template)"] = _.extend({
+    type: "(template)",
+    nud: doTemplateLiteral,
+    led: doTemplateLiteral,
+    noSubst: false
+  }, baseTemplateSyntax);
```
Comment (caitp) [inline src/jshint.js]: i like the metadata more than relying on just the "type" property. wdyt?

File: src/jshint.js [hunk 4]
```diff
@@ -2903,20 +2927,40 @@ var JSHINT = (function() {
...
-                 state.tokens.next.type !== "(end)") {
-        expression(10); // should probably have different rbp?
+  function doTemplateLiteral(left) {
+    // ASSERT: this.type === "(template)"
+    // jshint validthis: true
+    var ctx = this.context;
+    var noSubst = this.noSubst;
+    var depth = this.depth;
+
+    if (this === state.tokens.next) advance();
```
Comment (caitp) [inline src/jshint.js]: I don't think this check is needed anymore, since expression() will always use led/nud

File: src/jshint.js [hunk 5]
```diff
@@ -2903,20 +2927,40 @@ var JSHINT = (function() {
...
+        }
       }
     }
+
+    var tag = left;
     return {
       id: "(template)",
-      type: "(template)"
+      type: "(template)",
+      tag: tag
```
Comment (caitp) [inline src/jshint.js]: maybe we can do something useful with this at some point

File: src/jshint.js [hunk 6]
```diff
@@ -2903,20 +2927,40 @@ var JSHINT = (function() {
...
+        if (!state.tokens.next.template || state.tokens.next.depth > depth) {
+          expression(0); // should probably have different rbp?
+        } else {
+          // skip template start / middle
+          advance();
+        }
       }
     }
+
+    var tag = left;
```
Comment (caitp) [inline src/jshint.js]: admittedly, declaring this isn't particularly useful

File: src/jshint.js [hunk 7]
```diff
@@ -2903,20 +2927,40 @@ var JSHINT = (function() {
...
     return {
       id: "(template)",
-      type: "(template)"
+      type: "(template)",
+      tag: tag
     };
+
+    function end() {
+      if (state.tokens.curr.template && state.tokens.curr.tail &&
```
Comment (caitp) [inline src/jshint.js]: I know this is confusing, but there are a lot of things to juggle around u_U

File: src/lex.js [hunk 8]
```diff
@@ -129,7 +129,23 @@ Lexer.prototype = {
...
+
+  popContext: function() {
+    return this.context.pop();
+  },
+
+  isContext: function(context) {
+    return this.context.length > 0 && this.context[this.context.length - 1] === context;
+  },
+
+  currentContext: function() {
```
Comment (caitp) [inline src/lex.js]: currentContext() is probably not needed

File: src/lex.js [hunk 9]
```diff
@@ -1620,6 +1642,18 @@ Lexer.prototype = {
       if (token && token.startLine && token.startLine !== this.line) {
         obj.startLine = token.startLine;
       }
+      if (token && token.context) {
```
Comment (caitp) [inline src/lex.js]: there's a lot of extra metadata that the parser needs to use

**Comment by caitp** [comment]
did some self review, another set of eyes would be good though

File: src/lex.js (hunk 8 above)
Comment (caitp) [inline src/lex.js]: nevermind, I guess it is used

File: src/jshint.js [hunk 10]
```diff
@@ -904,7 +900,8 @@ var JSHINT = (function() {
         error("E030", state.tokens.curr, state.tokens.curr.id);
       }
 
-      while (rbp < state.tokens.next.lbp && !isEndOfExpr()) {
+      while ((rbp < state.tokens.next.lbp || state.tokens.next.type === "(template)") &&
```
Comment (caitp) [inline src/jshint.js]: todo: don't special case "(template)", just use the pratt mechanics for this

File: src/jshint.js (hunk 3 above)
Comment (rwaldron) [inline src/jshint.js]: I like it. Better metadata is always a win.

File: src/jshint.js (hunk 10 above)
Comment (rwaldron) [inline src/jshint.js]: Ping me when you're ready for another look

Comment (caitp) [inline src/jshint.js]: it's a TODO, i'm not sure how to make it fit into the pratt mechanics. I'll just add a comment marking it as such.

Comment (rwaldron) [inline src/jshint.js]: On Thu Feb 26 2015 at 12:17:05 PM ⭐caitp⭐ <notifications@github.com> wrote:

> In src/jshint.js
> <https://github.com/jshint/jshint/pull/2196#discussion_r25445080>:
>
> > @@ -904,7 +900,8 @@ var JSHINT = (function() {
> >          error("E030", state.tokens.curr, state.tokens.curr.id);
> >        }
> >
> > -      while (rbp < state.tokens.next.lbp && !isEndOfExpr()) {
> > +      while ((rbp < state.tokens.next.lbp || state.tokens.next.type === "(template)") &&
>
> it's a TODO, i'm not sure how to make it fit into the pratt mechanics.
> I'll just add a comment marking it as such.
>
Got it.

>

File: src/jshint.js [hunk 11]
```diff
@@ -1985,24 +1985,37 @@ var JSHINT = (function() {
...
+    noSubst: false
+  }, baseTemplateSyntax);
+
+  state.syntax["(no subst template)"] = _.extend({
+    type: "(template)",
+    nud: doTemplateLiteral,
+    led: doTemplateLiteral,
+    noSubst: true,
+    tail: true // mark as tail, since it's always the last component
+  }, baseTemplateSyntax);
```
Comment (rwaldron) [inline src/jshint.js]: This is really nice separation.

**Comment by ruimarinho** [comment]
@caitp, found a small issue and ticketed it as https://github.com/jshint/jshint/issues/2223. Thank you for adding this feature!

**Pull request merged**